### PR TITLE
Fix client crash in ResetServerGameSettings

### LINKED ISSUE #6921: Changing map can cause client to crash in `ResetServerGameSettings`
Copy a map (I tested with Aip-Gores) three times and keep switching between the copies on a LAN server with `sv_map`. After one or two changes the client will crash:

```
Error occurred on Wednesday, July 26, 2023 at 11:46:42.

DDNet.exe caused an Access Violation at location 00007FF7EBE03DC9 in module DDNet.exe Reading from location 0000000000000008.

AddrPC           Params
00007FF7EBE03DC9 00000213365CB040 000000000000001F 000000003F800000  DDNet.exe!CConsole::ResetServerGameSettings+0x87  [src/game/variables.h @ 194]
   192: MACRO_CONFIG_INT(SvVoteKickBantime, sv_vote_kick_bantime, 5, 0, 1440, CFGFLAG_SERVER, "The time in seconds to ban a player if kicked by vote. 0 makes it just use kick")
   193: MACRO_CONFIG_INT(SvJoinVoteDelay, sv_join_vote_delay, 300, 0, 1000, CFGFLAG_SERVER, "Add a delay before recently joined players can call any vote or participate in a kick/spec vote (in seconds)")
>  194: MACRO_CONFIG_INT(SvOldTeleportWeapons, sv_old_teleport_weapons, 0, 0, 1, CFGFLAG_SERVER | CFGFLAG_GAME, "Teleporting of all weapons (deprecated, use special entities instead)")
   195: MACRO_CONFIG_INT(SvOldTeleportHook, sv_old_teleport_hook, 0, 0, 1, CFGFLAG_SERVER | CFGFLAG_GAME, "Hook through teleporter (deprecated, use special entities instead)")
   196: MACRO_CONFIG_INT(SvTeleportHoldHook, sv_teleport_hold_hook, 0, 0, 1, CFGFLAG_SERVER | CFGFLAG_GAME, "Hold hook when teleported")
00007FF7EBCB008C 0000021335F65040 0000021300000000 0000000900000001  DDNet.exe!CGameClient::OnConnected+0x41a  [src/game/client/gameclient.cpp @ 525]
   523: mem_zero(&m_GameInfo, sizeof(m_GameInfo));
   524: m_PredictedDummyID = -1;
>  525: Console()->ResetServerGameSettings();
   526: LoadMapSettings();
   527: 
00007FF7EBB1ECF7 0000021335F65040 000000D1FB98BE90 0000021300000000  DDNet.exe!CClient::ProcessServerPacket+0xd07  [src/engine/client/client.cpp @ 1809]
  1807: else if(Conn == CONN_MAIN && (pPacket->m_Flags & NET_CHUNKFLAG_VITAL) != 0 && Msg == NETMSG_CON_READY)
  1808: {
> 1809: GameClient()->OnConnected();
  1810: }
  1811: else if(Conn == CONN_DUMMY && Msg == NETMSG_CON_READY)
00007FF7EBB21F58 0000021335F65040 0000021336802E70 0000021341C3EE40  DDNet.exe!CClient::PumpNetwork+0x38e  [src/engine/client/client.cpp @ 2569]
  2567: continue;
  2568: }
> 2569: ProcessServerPacket(&Packet, i, g_Config.m_ClDummy ^ i);
  2570: }
  2571: }
00007FF7EBB23565 0000021335F65040 000000D100000008 000000D1FB98CBFA  DDNet.exe!CClient::Update+0x11ad  [src/engine/client/client.cpp @ 2850]
  2848: 
  2849: // pump the network
> 2850: PumpNetwork();
  2851: 
  2852: if(m_pMapdownloadTask)
00007FF7EBB24CB7 000002133428B302 00007FF7EBB32A90 000000D1FB98EF83  DDNet.exe!CClient::Run+0xd73  [src/engine/client/client.cpp @ 3212]
  3210: }
  3211: 
> 3212: Update();
  3213: int64_t Now = time_get();
  3214: 
00007FF7EBB2E600 0000000000000001 0000000000000032 00007FF7EC1AFA18  DDNet.exe!main+0x10ee  [src/engine/client/client.cpp @ 4838]
  4836: // run the client
  4837: dbg_msg("client", "starting...");
> 4838: pClient->Run();
  4839: 
  4840: const bool Restarting = pClient->State() == CClient::STATE_RESTARTING;
```

At 22bd19be06b3adc2e6cf14550ce06c127239563a.

@sjrc6 I assume this is related to #6912.

### PR DESCRIPTION
**Pull request by sjrc6**
Fixes #6921

@Robyt3 Can you test this?

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [ ] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [ ] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Comment by sjrc6** [comment]
It seems very strange to me that this only happens if you spam map changes with sv_map. I have no idea what is going on.

**Comment by sjrc6** [comment]
To be fair the function ResetServerGameSettings() was never intended to be executed on the client, but there's no reason it can't be from what I can see. I am not very familiar with this part of the client though.

**Comment by heinrich5991** [comment]
> It seems very strange to me that this only happens if you spam map changes with sv_map. I have no idea what is going on.

That sounds like we should hold off on this PR until we start figuring out what happens.

**Comment by sjrc6** [comment]
> That sounds like we should hold off on this PR until we start figuring out what happens.

Yes that's fair. 

Also whats the deal with the failed check? I don't know why it's only coming up now.

**Comment by Robyt3** [comment]
The problem with the function `ResetServerGameSettings` is that it presumes that commands have a callback function and user data.

However, temporary commands which are registered for server commands when logged into rcon don't have a callback function and user data:

https://github.com/ddnet/ddnet/blob/22bd19be06b3adc2e6cf14550ce06c127239563a/src/engine/shared/console.cpp#L1146-L1147

The CI check fails because the `if` condition is added too late. Adding the `nullptr` check made clang-tidy aware that the variable can be `nullptr`, so it assumes that you pass a pointer to `nullptr` to `TraverseChain`, which would cause a crash there.

This can't be fixed by just ignore temporary commands (`pCommand->m_Temp`) in `ResetServerGameSettings`, as the temporary commands that don't have a callback function and user data seem to override the original non-temporary commands.

**Comment by sjrc6** [comment]
Do you have a proposed solution?

**Comment by heinrich5991** [comment]
I guess it's a bit weird that we register server commands to the client console. It might be best to revert #6912 for now.

**Comment by sjrc6** [comment]
@Robyt3 The client configs do not get over written by RegisterTemp, the issue is that FindCommand finds the temp ones before the "permanent" ones. This is easily fixed because the permanent ones use CFGFLAG_GAME | CFGFLAG_SERVER while the temp ones only use CFGFLAG_SERVER, so it's possible to filter for just the permanent ones.

**Review by Robyt3 — APPROVED** [review]
Thanks, I was going to ask if the temporary commands override the map settings.

**Comment by sjrc6** [comment]
Thanks for explaining the thing with the temp commands I would not have found that.

File: src/engine/shared/console.cpp [hunk 1]
```diff
@@ -1263,13 +1263,13 @@ const IConsole::CCommandInfo *CConsole::GetCommandInfo(const char *pName, int Fl
...
 
-void CConsole::ResetServerGameSettings()
+void CConsole::ResetServerGameSettings(int FlagMask)
 {
 #define MACRO_CONFIG_INT(Name, ScriptName, Def, Min, Max, Flags, Desc) \
 	{ \
 		if(((Flags) & (CFGFLAG_SERVER | CFGFLAG_GAME)) == (CFGFLAG_SERVER | CFGFLAG_GAME)) \
 		{ \
-			CCommand *pCommand = FindCommand(#ScriptName, CFGFLAG_SERVER); \
+			CCommand *pCommand = FindCommand(#ScriptName, FlagMask); \
```
Comment (heinrich5991) [inline src/engine/shared/console.cpp:1272]: This seems pretty cursed. It checks for `CFGFLAG_SERVER | CFGFLAG_GAME` being set above, and then uses the passed-in parameter to find the command? Perhaps it'd be better to just look for `CFGFLAG_GAME` and then find the command using `CFGFLAG_GAME`.

**Pull request merged**